source/simulator/core.cc

```diff
@@ -1767,15 +1767,53 @@ namespace aspect
     // start the principal loop over time steps:
     do
       {
-        // Only solve if we are not in pre-refinement, or we do not want to skip
-        // solving in pre-refinement.
+        // During pre-refinement, do not solve if we are asked to skip it:
         if (! (parameters.skip_solvers_on_initial_refinement
                && pre_refinement_step < parameters.initial_adaptive_refinement))
           {
             start_timestep ();
 
             // then do the core work: assemble systems and solve
-            solve_timestep ();
+            try
+              {
+                solve_timestep ();
+              }
+            catch (...)
+              {
+                pcout << "ERROR: the solver in the current timestep failed to converge." << std::endl;
+
+                if (parameters.nonlinear_solver_failure_strategy
+                    == Parameters<dim>::NonlinearSolverFailureStrategy::continue_with_next_timestep)
+                  {
+                    // do nothing and continue
+                  }
+                else if (parameters.nonlinear_solver_failure_strategy
+                         == Parameters<dim>::NonlinearSolverFailureStrategy::cut_timestep_size)
+                  {
+                    if (timestep_number == 0)
+                      {
+                        pcout << "Note: we can not cut the timestep in step 0, so we are aborting."
+                              << std::endl;
+                        throw;
+                      }
+
+                    // reduce timestep size and update time to "go back":
+                    const double new_time_step = 0.5*time_step;
+                    pcout << "\tReducing timestep to " << new_time_step << '\n' << std::endl;
+
+                    time -= time_step - new_time_step;
+                    time_step = new_time_step;
```

Comment:
There's one thing that I think doesn't work well with the current implementation: The time step size is reduced for the current time step, until the solver converges, but the next time step bases the time step size on the CFL number (rather than on what time step size was necessary to actually get the nonlinear solver to converge). So if the problem stays as hard to solve, we need to go though a number of failing nonlinear solver loops for every time step, rather than continuing with the new (smaller) time step that is adapted to the nonlinear problem.

Reply:
That makes sense, but how do you want to decide when to make bigger timestep again? I don't think there is a good answer here. One thing that might help in an extreme case is not increasing it by more than a factor of say 2 in each step. But in general, this seems to be something that would require more user input.